File: Implement Queue using Stacks/Implement_Queue_using_Stacks.py
# Runtime: 36 ms, faster than 50.15% of Python3 online submissions for Implement Queue using Stacks.
# Memory Usage: 13.1 MB, less than 5.32% of Python3 online submissions for Implement Queue using Stacks.

class MyQueue:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.stack1 = []
        self.stack2 = []
        self.topElement = None

    def push(self, x: int) -> None:
        """
        Push element x to the back of queue.
        """
        if self.empty():
            self.topElement = x
        elif not self.stack1:
            while self.stack2:
                self.stack1.append(self.stack2.pop())
        self.stack1.append(x)
        

    def pop(self) -> int:
        """
        Removes the element from in front of queue and returns that element.
        """
        if self.stack2:
            res = self.stack2.pop()
            if self.stack2:
                self.topElement = self.stack2[-1]
            return res
        else:
            temp = self.stack1[-1]
            while len(self.stack1) != 1:
                temp = self.stack1.pop()
                self.stack2.append(temp)
            
            self.topElement = temp

            return self.stack1.pop()
        

    def peek(self) -> int:
        """
        Get the front element.
        """
        return self.topElement
        

    def empty(self) -> bool:
        """
        Returns whether the queue is empty.
        """
        return True if not self.stack1 and not self.stack2 else False
            
# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()

File: Implement Queue using Stacks/Implement_Queue_using_Stacks.cpp
// 基本思路 使用两个堆栈来模拟队列

// Runtime: 4 ms, faster than 100.00% of C++ online submissions for Implement Queue using Stacks.
// Memory Usage: 9 MB, less than 49.26% of C++ online submissions for Implement Queue using Stacks.

class MyQueue 
{
public:
    /** Initialize your data structure here. */
    MyQueue() 
    {
        
    }
    
    /** Push element x to the back of queue. */
    void push(int x) 
    {
        if (empty())
            topElement = x;

        else if (stack1.empty())
        {
            topElement = stack2.top();
            while (!stack2.empty())
            {
                int temp = stack2.top();
                stack2.pop();
                stack1.push(temp);
            }
        }

        stack1.push(x);
    }
    
    /** Removes the element from in front of queue and returns that element. */
    int pop() 
    {
        if (!stack2.empty())
        {
            int res = stack2.top();
            stack2.pop();

            if (!stack2.empty())
                topElement = stack2.top();
            return res;
        }
        else
        {
            int temp;
            while (stack1.size() != 1)
            {
                temp = stack1.top();
                stack1.pop();
                stack2.push(temp)
            }

            topElement = temp;

            int res = stack1.top();
            stack1.pop();
            return res;
        }
        
    }
    
    /** Get the front element. */
    int peek() 
    {
        return topElement;
    }
    
    /** Returns whether the queue is empty. */
    bool empty() 
    {
        if (stack1.empty() && stack2.empty())
            return true;
        else
            return false;
    }

private:
    stack<int> stack1, stack2;
    int topElement;
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * bool param_4 = obj.empty();
 */